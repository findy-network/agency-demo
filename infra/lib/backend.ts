import { Construct } from 'constructs'

import * as lightsail from 'aws-cdk-lib/aws-lightsail'

interface BackendProps {
  readonly account: string
  readonly region: string
}

export class Backend extends Construct {
  public readonly appUrl: string

  constructor(scope: Construct, id: string, props: BackendProps) {
    super(scope, id)

    new lightsail.CfnContainer(scope, 'FindyAgencyDemoBackend', {
      scale: 1,
      power: 'nano',
      serviceName: 'agency-demo',
      containerServiceDeployment: {
        containers: [
          {
            containerName: 'agency-demo',
            image: `ghcr.io/findy-network/agency-demo:latest`,
            ports: [{ port: '5000', protocol: 'HTTP' }],
            environment: [
              {
                variable: 'AGENCY_AUTH_URL',
                value: process.env.BACKEND_AGENCY_AUTH_URL,
              },
              {
                variable: 'AGENCY_AUTH_ORIGIN',
                value: process.env.BACKEND_AGENCY_AUTH_ORIGIN,
              },
              {
                variable: 'AGENCY_USER_NAME',
                value: process.env.BACKEND_AGENCY_USER_NAME,
              },
              {
                variable: 'AGENCY_PUBLIC_DID_SEED',
                value: process.env.BACKEND_AGENCY_PUBLIC_DID_SEED,
              },
              {
                variable: 'AGENCY_KEY',
                value: process.env.BACKEND_AGENCY_KEY,
              },
              {
                variable: 'SERVER_ADDRESS',
                value: process.env.BACKEND_SERVER_ADDRESS,
              },
              {
                variable: 'SERVER_PORT',
                value: process.env.BACKEND_SERVER_PORT,
              },
              {
                variable: 'WALLET_URL',
                value: process.env.BACKEND_WALLET_URL,
              },
            ],
          },
        ],
        publicEndpoint: {
          containerName: 'agency-demo',
          containerPort: 5000,
        },
      },
    })

    this.appUrl = (process.env.BACKEND_URL || 'example.com').replace('https://', '').replace('/', '') // TODO
  }
}
