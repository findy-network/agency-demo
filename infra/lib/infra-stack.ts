import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";

import { Frontend } from "./frontend";
import { Backend } from "./backend";


export class InfraStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const backend = new Backend(this, `${id}-backend`, {
      account: this.account,
      region: this.region
    });

    const apiPaths = [
      "/agent*",
      "/oob*",
      "/connections*",
      "/credentials*",
      "/proofs*",
      "/demo*",
      "/server*",
      "/public*",
      "/ws*",
    ];
    const backendUrl = backend.appUrl;
    new Frontend(this, `${id}-frontend`, {
      rootDomainName: process.env.DOMAIN_NAME || "example.com",
      appDomainPrefix: process.env.SUB_DOMAIN_NAME || "example",
      apiPaths,
      backendUrl,
    });
  }
}
