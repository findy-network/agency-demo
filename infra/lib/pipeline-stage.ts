import { InfraStack } from './infra-stack'
import { Stage, StageProps } from 'aws-cdk-lib'
import { Construct } from 'constructs'

export class InfraPipelineStage extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props)

    new InfraStack(this, 'FindyAgencyDemoInfraStack', props)
  }
}
