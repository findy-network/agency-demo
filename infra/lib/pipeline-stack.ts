import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import {
    CodeBuildStep,
    CodePipeline,
    CodePipelineSource,
} from "aws-cdk-lib/pipelines";
import { aws_codebuild as codebuild, aws_logs as logs } from "aws-cdk-lib";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";

import { InfraPipelineStage } from "./pipeline-stage";
import { NotificationRule } from "aws-cdk-lib/aws-codestarnotifications";
import { Topic } from "aws-cdk-lib/aws-sns";

interface InfraPipelineProperties extends cdk.StackProps {
}

const environmentVariables: Record<string, codebuild.BuildEnvironmentVariable> =
{
    DOMAIN_NAME: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/domain-name",
    },
    SUB_DOMAIN_NAME: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/sub-domain-name",
    },
    // TODO: use secrets
    BACKEND_AGENCY_USER_NAME: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-agency-user-name"
    },
    // TODO: use secrets
    BACKEND_AGENCY_KEY: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-agency-key"
    },
    BACKEND_SERVER_PORT: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-server-port"
    },
    BACKEND_AGENCY_AUTH_URL: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-agency-auth-url"
    },
    BACKEND_AGENCY_AUTH_ORIGIN: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-agency-auth-origin"
    },
    // TODO: use secrets
    BACKEND_AGENCY_PUBLIC_DID_SEED: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-agency-public-did-seed"
    },
    BACKEND_SERVER_ADDRESS: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-server-address"
    },
    BACKEND_WALLET_URL: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-wallet-url"
    },
    BACKEND_URL: {
        type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        value: "/agency-demo/backend-url"
    },
};

export class InfraPipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props: InfraPipelineProperties) {
        super(scope, id, props);

        const githubConnectionArn = StringParameter.valueForStringParameter(
            this,
            "/agency-demo/github-connection-arn"
        );
        const source = CodePipelineSource.connection(
            "findy-network/agency-demo",
            "master",
            {
                connectionArn: githubConnectionArn, // Created using the AWS console
            }
        )

        // Create pipeline
        const pipeline = this.createPipeline(source);

        // Add app to pipeline
        const deploy = new InfraPipelineStage(this, "Deploy", {
            env: props.env,
        });
        const deployStage = pipeline.addStage(deploy);

        // Add lightsail application update step
        deployStage.addPost(this.createLightsailUpdateStep(id));

        // Add frontend build step
        const frontBuildStep = this.createFrontendBuildStep();
        deployStage.addPost(frontBuildStep);

        // Add frontend deploy step
        const frontDeployStep = this.createFrontendDeployStep(frontBuildStep);
        deployStage.addPost(frontDeployStep);

        // Add deployment test step
        //deployStage.addPost(this.createPostDeploymentTestStep(frontDeployStep, source));

        // need this to add the notification rule
        pipeline.buildPipeline();

        new NotificationRule(this, "FindyAgencyDemoPipelineNotificationRule", {
            source: pipeline.pipeline,
            events: [
                "codepipeline-pipeline-pipeline-execution-failed",
                "codepipeline-pipeline-pipeline-execution-canceled",
                "codepipeline-pipeline-pipeline-execution-started",
                "codepipeline-pipeline-pipeline-execution-resumed",
                "codepipeline-pipeline-pipeline-execution-succeeded",
            ],
            targets: [new Topic(this, "FindyAgencyDemoPipelineNotificationTopic")],
        });

        // manually adjust logs retention
        // this.node.findAll().forEach((construct, index) => {
        //     if (construct instanceof codebuild.Project) {
        //         new logs.LogRetention(this, `LogRetention${index}`, {
        //             logGroupName: `/aws/codebuild/${construct.projectName}`,
        //             retention: logs.RetentionDays.ONE_MONTH,
        //         });
        //     }
        // });
    }

    createPipeline(source: CodePipelineSource) {
        const pipeline = new CodePipeline(this, "Pipeline", {
            pipelineName: "FindyAgencyDemoPipeline",
            synth: new CodeBuildStep("SynthStep", {
                input: source,
                installCommands: ["npm install -g aws-cdk"],
                buildEnvironment: {
                    environmentVariables: {
                        CDK_CONTEXT_JSON: {
                            type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
                            value: "/agency-demo/cdk-context",
                        },
                    },
                },
                commands: [
                    "cd infra",
                    `echo "$CDK_CONTEXT_JSON" > cdk.context.json`,
                    "cat cdk.context.json",
                    "npm ci",
                    "npm run build",
                    "npx cdk synth",
                    "npm run pipeline:context",
                ],
                rolePolicyStatements: [
                    new PolicyStatement({
                        actions: ["ssm:PutParameter"],
                        resources: [
                            `arn:aws:ssm:${this.region}:${this.account}:parameter/agency-demo*`,
                        ],
                    }),
                ],
                primaryOutputDirectory: "infra/cdk.out",
            }),
            codeBuildDefaults: {
                buildEnvironment: {
                    environmentVariables: {
                        ...environmentVariables,
                    },
                },
            },
        });

        return pipeline;
    }

    createLightsailUpdateStep(id: string) {
        const deployRole = new Role(this, `${id}-deploy-role`, {
            assumedBy: new ServicePrincipal("codebuild.amazonaws.com"),
        });

        const lightsailPolicy = new PolicyStatement();
        lightsailPolicy.addActions(
            "lightsail:*",
            "ssm:*"
        );
        lightsailPolicy.addResources("*");
        deployRole.addToPolicy(lightsailPolicy);

        return new CodeBuildStep("FindyAgencyDemoDeployBackendStep", {
            projectName: "FindyAgencyDemoDeployBackendStep",
            commands: [
                //`VERSION="$(./tools/version.sh ./)-$(date +%s)"`,
                //`aws lightsail update-container-service --service-name "agency-demo"`,
                `URL=$(aws lightsail get-container-services --service-name agency-demo --output json | jq '.containerServices[0].url')`,
                `aws ssm put-parameter --overwrite --name \"/agency-demo/backend-url\" --value \"$URL\" --type String`
            ],
            role: deployRole,
        });
    }

    createFrontendBuildStep() {
        return new CodeBuildStep("FindyAgencyDemoBuildFrontendStep", {
            projectName: "FindyAgencyDemoBuildFrontend",
            commands: [
                "apk add bash",
                "yarn install",
                "yarn client:build"
            ],
            buildEnvironment: {
                buildImage: codebuild.LinuxBuildImage.fromDockerRegistry(
                    "node:16.18-alpine3.16"
                ),
                environmentVariables: {
                    REACT_APP_HOST_BACKEND: {
                        value: `https://${process.env.SUB_DOMAIN_NAME}.${process.env.DOMAIN_NAME}`,
                    },
                    REACT_APP_HOST_WEBSOCKET: {
                        value: `wss://${process.env.SUB_DOMAIN_NAME}.${process.env.DOMAIN_NAME}`,
                    },
                },
            },
            primaryOutputDirectory: "client/build",
        });
    }

    createFrontendDeployStep(frontBuildStep: CodeBuildStep) {
        return new CodeBuildStep("FindyAgencyDemoDeployFrontendStep", {
            input: frontBuildStep.primaryOutput,
            projectName: "FindyAgencyDemoDeployFrontend",
            commands: [
                `V1=$(curl https://$SUB_DOMAIN_NAME.$DOMAIN_NAME/version.txt)`,
                `V2=$(cat ./version.txt)`,
                `if [ "$V1" != "$V2" ]; then aws s3 sync --delete . s3://$SUB_DOMAIN_NAME.$DOMAIN_NAME; fi`,
            ],
            rolePolicyStatements: [
                new PolicyStatement({
                    actions: ["s3:Put*", "s3:Delete*", "s3:Get*", "s3:List*"],
                    resources: ["*"],
                }),
            ],
            primaryOutputDirectory: ".",
        });
    }

    // createPostDeploymentTestStep(frontDeployStep: CodeBuildStep, source: CodePipelineSource) {
    //     return new CodeBuildStep("FindyAgencyDemoDeploymentTest", {
    //         input: source,
    //         additionalInputs: { out: frontDeployStep.primaryOutput! },
    //         projectName: "FindyAgencyDemoDeploymentTest",
    //         // TODO: make more extensive tests
    //         commands: [
    //             `curl -sI https://$SUB_DOMAIN_NAME.$DOMAIN_NAME/version.txt | grep "HTTP/2 200"`,
    //             `./infra/tools/wait-for-ready.sh`,
    //         ],
    //     });
    // }
}
