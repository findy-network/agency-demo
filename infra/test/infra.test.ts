import * as cdk from 'aws-cdk-lib'
import { Stack } from 'aws-cdk-lib'
import { Template, Match } from 'aws-cdk-lib/assertions'
import * as Infra from '../lib/infra-stack'

test('Infra Stack Created', () => {
  const app = new cdk.App()
  const stack = new Infra.InfraStack(app, 'MyTestStack', {
    env: { account: '123456789012', region: 'us-east-1' },
  })
  const template = Template.fromStack(stack)

  expect(template).toMatchSnapshot()

  template.resourceCountIs('AWS::Lightsail::Container', 1)
  template.resourceCountIs('AWS::CloudFront::Distribution', 1)
  template.resourceCountIs('AWS::CloudFront::CloudFrontOriginAccessIdentity', 1)
  template.resourceCountIs('AWS::S3::Bucket', 1)
  template.hasResourceProperties('AWS::S3::Bucket', {
    BucketName: 'example.example.com',
    PublicAccessBlockConfiguration: {
      BlockPublicAcls: true,
      BlockPublicPolicy: true,
      IgnorePublicAcls: true,
      RestrictPublicBuckets: true,
    },
  })
  template.hasResourceProperties('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Aliases: ['example.example.com'],
    },
  })
})
