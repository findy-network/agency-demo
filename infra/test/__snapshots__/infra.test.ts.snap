// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infra Stack Created 1`] = `
Object {
  "Mappings": Object {
    "AWSCloudFrontPartitionHostedZoneIdMap": Object {
      "aws": Object {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": Object {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "FindyAgencyDemoBackend": Object {
      "Properties": Object {
        "ContainerServiceDeployment": Object {
          "Containers": Array [
            Object {
              "ContainerName": "agency-demo",
              "Environment": Array [
                Object {
                  "Variable": "AGENCY_AUTH_URL",
                },
                Object {
                  "Variable": "AGENCY_AUTH_ORIGIN",
                },
                Object {
                  "Variable": "AGENCY_USER_NAME",
                },
                Object {
                  "Variable": "AGENCY_PUBLIC_DID_SEED",
                },
                Object {
                  "Variable": "AGENCY_KEY",
                },
                Object {
                  "Variable": "SERVER_ADDRESS",
                },
                Object {
                  "Variable": "SERVER_PORT",
                },
                Object {
                  "Variable": "WALLET_URL",
                },
              ],
              "Image": "ghcr.io/findy-network/agency-demo:latest",
              "Ports": Array [
                Object {
                  "Port": "5000",
                  "Protocol": "HTTP",
                },
              ],
            },
          ],
          "PublicEndpoint": Object {
            "ContainerName": "agency-demo",
            "ContainerPort": 5000,
          },
        },
        "Power": "nano",
        "Scale": 1,
        "ServiceName": "agency-demo",
      },
      "Type": "AWS::Lightsail::Container",
    },
    "MyTestStackfrontendMyTestStackfrontendarecordFA69DF2A": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "MyTestStackfrontendMyTestStackfrontenddistributionCFDistributionCE45D98B",
              "DomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::FindInMap": Array [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              Object {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "example.example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MyTestStackfrontendMyTestStackfrontendbucket32024486": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "example.example.com",
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "WebsiteConfiguration": Object {
          "ErrorDocument": "index.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestStackfrontendMyTestStackfrontendbucketPolicy198AA21E": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestStackfrontendMyTestStackfrontendbucket32024486",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject*",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackfrontendMyTestStackfrontendoriginaccessidentityCBE7DCDB",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "MyTestStackfrontendMyTestStackfrontendbucket32024486",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:ListBucket",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackfrontendMyTestStackfrontendoriginaccessidentityCBE7DCDB",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MyTestStackfrontendMyTestStackfrontendbucket32024486",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestStackfrontendMyTestStackfrontendoriginaccessidentityCBE7DCDB",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "MyTestStackfrontendMyTestStackfrontendbucket32024486",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunction2A30CDB8": Object {
      "DependsOn": Array [
        "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRoleDefaultPolicyD8E1CDCE",
        "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRole54B6A1AA",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "861509f9607aa6bb9fd7034fb10fe6f0c60d91dfd7f010b9ef666869dec9bbd9.zip",
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRole54B6A1AA",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRole54B6A1AA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRoleDefaultPolicyD8E1CDCE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Condition": Object {
                "ForAllValues:StringEquals": Object {
                  "route53:ChangeResourceRecordSetsActions": Array [
                    "UPSERT",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": Array [
                    "CNAME",
                  ],
                },
                "ForAllValues:StringLike": Object {
                  "route53:ChangeResourceRecordSetsNormalizedRecordNames": Array [
                    "*.example.example.com",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/DUMMY",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRoleDefaultPolicyD8E1CDCE",
        "Roles": Array [
          Object {
            "Ref": "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunctionServiceRole54B6A1AA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorResourceBC6C0DEA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "example.example.com",
        "HostedZoneId": "DUMMY",
        "Region": "us-east-1",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorFunction2A30CDB8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestStackfrontendMyTestStackfrontenddistributionCFDistributionCE45D98B": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.example.com",
          ],
          "CacheBehaviors": Array [
            Object {
              "AllowedMethods": Array [
                "GET",
                "HEAD",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "DefaultTTL": 0,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "none",
                },
                "QueryString": false,
              },
              "MaxTTL": 0,
              "MinTTL": 0,
              "PathPattern": "/index.html",
              "TargetOriginId": "origin1",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "GET",
                "HEAD",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "DefaultTTL": 0,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "none",
                },
                "QueryString": false,
              },
              "MaxTTL": 0,
              "MinTTL": 0,
              "PathPattern": "/version.txt",
              "TargetOriginId": "origin1",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/oob/*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/connections*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/credentials/*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/proofs/*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/demo/*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/server/last-reset",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/public/*",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
            Object {
              "AllowedMethods": Array [
                "DELETE",
                "GET",
                "HEAD",
                "OPTIONS",
                "PATCH",
                "POST",
                "PUT",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "all",
                },
                "Headers": Array [
                  "*",
                ],
                "QueryString": true,
              },
              "PathPattern": "/ws",
              "TargetOriginId": "origin2",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
          ],
          "CustomErrorResponses": Array [
            Object {
              "ErrorCachingMinTTL": 0,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "MyTestStackfrontendMyTestStackfrontendbucket32024486",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestStackfrontendMyTestStackfrontendoriginaccessidentityCBE7DCDB",
                      },
                    ],
                  ],
                },
              },
            },
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 60,
                "OriginProtocolPolicy": "https-only",
                "OriginReadTimeout": 60,
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": "example.com",
              "Id": "origin2",
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": Object {
              "Fn::GetAtt": Array [
                "MyTestStackfrontendMyTestStackfrontendcertificateCertificateRequestorResourceBC6C0DEA",
                "Arn",
              ],
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestStackfrontendMyTestStackfrontendoriginaccessidentityCBE7DCDB": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Access bucket example.example.com only from Cloudfront",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
